Source/WebCore/platform/graphics/ca/TileController.cpp

```diff
@@ -116,6 +116,114 @@ std::optional<TileGridIdentifier> TileController::secondaryGridIdentifier() cons
     return { };
 }
 
+std::optional<TileConfigurationChangeIdentifier> TileController::activeConfigurationChange() const
+{
+    return m_asyncTileConfigurationChangeData.transform([](const auto& data) {
+        return data.tileConfigurationChange;
+    });
+}
+
+std::optional<TileGridIdentifier> TileController::pendingGridForTileConfigurationChange(const std::optional<TileConfigurationChangeIdentifier>& tileConfigurationChange) const
+{
+    return tileConfigurationChange.and_then([&changeData = m_asyncTileConfigurationChangeData](const auto& change) {
+        return changeData.and_then([&change](const auto& data) -> std::optional<TileGridIdentifier> {
+            if (data.tileConfigurationChange != change)
+                return std::nullopt;
+            return data.pendingTileGrid->identifier();
+        });
+    });
+}
+
+void TileController::destroyExistingTileConfigurationChange()
+{
+    ASSERT(m_asyncTileConfigurationChangeData);
+
+    auto tileConfigurationChange = m_asyncTileConfigurationChangeData->tileConfigurationChange;
+    if (m_client)
+        m_client->didCancelTileConfigurationChange(tileConfigurationChange);
+
+    /*
+     // FIXME: Maybe we want to uncnoditionally DidCancelTileConfigurationChange::No here.
+     auto didCancelTileConfigurationChange = m_asyncTileConfigurationChangeData->pendingTiles.isEmpty() ? DidCancelTileConfigurationChange::No : DidCancelTileConfigurationChange::Yes;
+     m_asyncTileConfigurationChangeData->completionHandler(primaryGridIdentifier(), m_asyncTileConfigurationChangeData->pendingTileGrid->identifier(), didCancelTileConfigurationChange);
+     */
+
+    m_asyncTileConfigurationChangeData->completionHandler(primaryGridIdentifier(), { });
+    m_asyncTileConfigurationChangeData.reset();
+}
+
+void TileController::addPendingTilesToActiveTileConfigurationChangeIfNeeded(const TileGridIndexIteratorRange& tiles)
+{
+    if (!m_asyncTileConfigurationChangeData)
+        return;
+
+    for (const auto& tile : tiles) {
+        auto newTile = *m_asyncTileConfigurationChangeData->pendingTiles.add(tile).iterator;
+        auto tileRect = m_asyncTileConfigurationChangeData->pendingTileGrid->rectForTile(newTile);
+        if (m_client) {
+            m_client->prepareContentForTile(*this, m_asyncTileConfigurationChangeData->pendingTileGrid->identifier(), newTile, tileRect, *activeConfigurationChange(), [newTile, this](auto completedTile) {
+                ASSERT_UNUSED(newTile, newTile == completedTile);
+                didPrepareContentForTile(completedTile);
+            });
+        }
+    }
+}
+
+TileConfigurationChangeIdentifier TileController::initiateTileConfigurationChange(CompletionHandler<void(TileGridIdentifier oldGrid, const std::optional<TileGridIdentifier>& newGrid)>&& completionHandler)
+{
+    if (m_asyncTileConfigurationChangeData)
+        destroyExistingTileConfigurationChange();
```

Comment:
This seems potentially problematic.

If we're pinch zooming the PDF, and new scales are arriving to the main thread continuously at a rate faster than the plugin can paint them, won't this just cancel all of the paints?

I think that'd mean you'd just have the initial tiles displayed the whole tile (stretched to the new scale), while we keep doing lots of painting work and not displaying it.

If anything else inside the plugin is changing rendering at the same time (like an animation or something), then those changes wouldn't get displayed until you stopped pinch zooming.

It seems like we should probably suppress scale changes from reaching the plugin/tilecontroller if an existing configuration change is in progress. We'd then need to wait for the commit, the following rendering update (which is where the committed new tile grid gets serialised into a RemoteLayerTreeTransaction), and then schedule yet another rendering update. At that point we can pick up the new scale, and initiate a new configuration change.